File: matiasmasso/Mat.NET/DT2/DTOCur.vb
Public Class DTOCur
    Property tag As String
    Property nom As DTOLangText
    Property symbol As String
    Property decimals As Integer = 2
    Property exchangeRate As DTOCurExchangeRate
    Property obsoleto As Boolean


    Public Enum Ids
        NotSet
        EUR
        USD
        GBP
        ESP
    End Enum

    Public Sub New()
        MyBase.New
    End Sub

    Shared Function Factory(sTag As String) As DTOCur
        Return DTOApp.Current.Curs.FirstOrDefault(Function(x) x.Tag = sTag.ToUpper)
    End Function


    Public Function Clon() As DTOCur
        Dim retval = DTOCur.Factory(_Tag)
        With retval
            .Nom = _Nom
            .Symbol = _Symbol
            .Decimals = _Decimals
            .ExchangeRate = _ExchangeRate
            .Obsoleto = _Obsoleto
        End With
        Return retval
    End Function

    Shared Function NomOrTag(oCur As DTOCur, oLang As DTOLang) As String
        Dim retval As String = oCur.Nom.Tradueix(oLang)
        If retval = "" Then retval = oCur.Tag
        Return retval
    End Function


    Shared Function Eur() As DTOCur
        Return DTOCur.Factory(DTOCur.Ids.EUR.ToString)
    End Function

    Shared Function Usd() As DTOCur
        Return DTOCur.Factory(DTOCur.Ids.USD.ToString)
    End Function

    Shared Function Gbp() As DTOCur
        Return DTOCur.Factory(DTOCur.Ids.GBP.ToString)
    End Function

    Public Function FormatString() As String
        Dim sb As New System.Text.StringBuilder
        sb.Append("#,##0")
        If _Decimals > 0 Then sb.Append("." & New String("0", _Decimals))
        If _Symbol > "" Then
            sb.Append(" " & _Symbol)
        Else
            sb.Append(" " & _Tag)
        End If
        sb.Append(";-" & sb.ToString & ";#")
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Public Shadows Function Equals(ByVal oCur As DTOCur) As Boolean
        Dim retVal As Boolean = False
        If oCur IsNot Nothing Then
            If _Tag = oCur.Tag Then retVal = True
        End If
        Return retVal
    End Function

    Public Shadows Function UnEquals(ByVal oCur As DTOCur) As Boolean
        Dim retVal As Boolean = Not Equals(oCur)
        Return retVal
    End Function


    Public Function AmtFromEuros(Euros As Decimal, Optional oExchangeRate As DTOCurExchangeRate = Nothing) As DTOAmt
        If oExchangeRate Is Nothing Then oExchangeRate = _ExchangeRate
        Dim Val As Decimal = Math.Round(Euros * oExchangeRate.Rate, 2, MidpointRounding.AwayFromZero)
        Dim retval = DTOAmt.Factory(Euros, Me, Val)
        Return retval
    End Function

    Public Function ExchangeText(Optional oExchangeRate As DTOCurExchangeRate = Nothing) As String
        If oExchangeRate Is Nothing Then oExchangeRate = _ExchangeRate
        Dim retval As String = String.Format("{0} {1}/€ {2:dd/MM/yy}", oExchangeRate.Rate, _Symbol, oExchangeRate.Fch)
        Return retval
    End Function


    Public Function AmtFromDivisa(Divisa As Decimal, Optional oExchangeRate As DTOCurExchangeRate = Nothing) As DTOAmt
        If oExchangeRate Is Nothing Then oExchangeRate = _ExchangeRate
        Dim Eur As Decimal = Math.Round(Divisa / oExchangeRate.Rate, 2, MidpointRounding.AwayFromZero)
        Dim retval = DTOAmt.Factory(Eur, Me, Divisa)
        Return retval
    End Function


    Public Function isEur() As Boolean
        Return _Tag = "EUR"
    End Function

    Public Function isUSD() As Boolean
        Return _Tag = "USD"
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOCur.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOCur
    {
        public string Tag { get; set; }
        public string Symbol { get; set; }
        public int Decimals { get; set; } = 2;
        public DTOCurExchangeRate ExchangeRate { get; set; }
        public bool Obsoleto { get; set; }


        public enum Ids
        {
            NotSet,
            EUR,
            USD,
            GBP,
            ESP
        }

        public DTOCur() : base()
        {
        }

        public static DTOCur Factory(string sTag)
        {
            DTOCur retval = null;
            if (DTOApp.Current == null || DTOApp.Current.Curs == null) // for user controls in design mode
            {
                retval = new DTOCur();
                retval.Tag = "EUR";
            }
            else

                retval = DTOApp.Current.Curs.FirstOrDefault(x => x.Tag == sTag.ToUpper());
            return retval;
        }


        public DTOCur clon()
        {
            var retval = DTOCur.Factory(Tag);
            {
                var withBlock = retval;
                withBlock.Symbol = Symbol;
                withBlock.Decimals = Decimals;
                withBlock.ExchangeRate = ExchangeRate;
                withBlock.Obsoleto = Obsoleto;
            }
            return retval;
        }

        public static DTOCur Eur()
        {
            return DTOCur.Factory(DTOCur.Ids.EUR.ToString());
        }

        public static DTOCur usd()
        {
            return DTOCur.Factory(DTOCur.Ids.USD.ToString());
        }

        public static DTOCur gbp()
        {
            return DTOCur.Factory(DTOCur.Ids.GBP.ToString());
        }

        public string formatString()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append("#,##0");
            if (Decimals > 0)
                sb.Append("." + string.Concat(Enumerable.Repeat("0", Decimals)));
            if (Symbol.isNotEmpty())
                sb.Append(" " + Symbol);
            else if (Tag == "EUR")
                sb.Append(" €");
            else
                sb.Append(" " + Tag);
            sb.Append(";-" + sb.ToString() + ";#");
            string retval = sb.ToString();
            return retval;
        }


        public bool UnEquals(DTOCur oCur)
        {
            bool retVal = Tag != oCur.Tag;
            return retVal;
        }


        public DTOAmt amtFromEuros(decimal euros, DTOCurExchangeRate oExchangeRate = null/* TODO Change to default(_) if this is not a reference type */)
        {
            if (oExchangeRate == null)
                oExchangeRate = ExchangeRate;
            decimal Val = Math.Round(euros * oExchangeRate.Rate, 2, MidpointRounding.AwayFromZero);
            var retval = DTOAmt.Factory(euros, this, Val);
            return retval;
        }

        public string exchangeText(DTOCurExchangeRate oExchangeRate = null/* TODO Change to default(_) if this is not a reference type */)
        {
            if (oExchangeRate == null)
                oExchangeRate = ExchangeRate;
            string retval = string.Format("{0} {1}/€ {2:dd/MM/yy}", oExchangeRate.Rate, Symbol, oExchangeRate.Fch);
            return retval;
        }


        public DTOAmt amtFromDivisa(decimal divisa, DTOCurExchangeRate oExchangeRate = null/* TODO Change to default(_) if this is not a reference type */)
        {
            if (oExchangeRate == null)
                oExchangeRate = ExchangeRate;
            decimal eur = Math.Round(divisa / (decimal)oExchangeRate.Rate, 2, MidpointRounding.AwayFromZero);
            var retval = DTOAmt.Factory(eur, this, divisa);
            return retval;
        }


        public bool isEur()
        {
            return Tag == "EUR";
        }

        public bool isUSD()
        {
            return Tag == "USD";
        }

        public override string ToString()
        {
            return Tag;
        }

        public class Compact
        {
            public string Tag { get; set; }

            public static Compact Factory(string tag)
            {
                Compact retval = new Compact();
                retval.Tag = tag;
                return retval;
            }

        }

        public class Collection : List<DTOCur>
        {
            public Collection() : base()
            {

            }
        }
    }
}
